### ISSUE TITLE
Different behaviors between Jupyter Notebook and VS Code in StructuredGrid.plot()

### ISSUE THREAD

**Issue by leomiquelutti** [pyvista/pyvista-support#206]
Hi everyone,

## 1 - StructuredGrid.plot() in Jupyter Notebook and VS Code

I am running both examples of [create-structured-surface](https://docs.pyvista.org/examples/00-load/create-structured-surface.html#sphx-glr-examples-00-load-create-structured-surface-py) provided in the page with:
- Jupyter Notebook - the .ipybn file
- Visual Studio Code (VS Code) - the .py file

However, I get different results for the same parts of the code. VS Code makes it white, while Jupyter Notebook makes it colored. The piece of code that corresponds to the images is:

```
import pyvista as pv
from pyvista import examples
import numpy as np

###############################
# Create a simple meshgrid using NumPy

# Make data
x = np.arange(-10, 10, 0.25)
y = np.arange(-10, 10, 0.25)
x, y = np.meshgrid(x, y)
r = np.sqrt(x ** 2 + y ** 2)
z = np.sin(r)

#################################
# Now pass the NumPy meshgrid to PyVista

# Create and plot structured grid
grid = pv.StructuredGrid(x, y, z)
grid.plot(cmap='jet')
```
![Anotação 2020-07-14 111743](https://user-images.githubusercontent.com/9113903/87440646-9577aa00-c5c8-11ea-8c11-2b8a797cfa9d.png)

It seems there something about `StructuredGrid.plot()` in VS Code. Whether I pass or don't `cmap` as an argument, it does not make any difference. I tried two different environments:
* from [gostin](https://github.com/softwareunderground/gostin)
* following [the official guide](https://docs.pyvista.org/getting-started/installation.html)

## 2 - My own data in Jupyter Notebook and VS Code

However, it gets worse!

When I run this code below (data available [here](https://1drv.ms/u/s!AvWFfSwwsz6gloRk8c9mm9gF8-EUXA?e=oyAf14)) either in Jupyter Notebook or VS Code, the output also gets white, whether I use `StructuredGrid.plot()` or `Plotter.show()`

```
import pyvista as pv

# read StructuredGrid data
pvgrid = pv.read('my_pv_grid.vtk')

# StructuredGrid.plot()
pvgrid.plot()

# Plotter.show()
p = pv.Plotter(notebook=False)
p.add_mesh(pvgrid)
p.show()
```
![image](https://user-images.githubusercontent.com/9113903/87442336-901b5f00-c5ca-11ea-8a7b-21ab75d2ba41.png)

## 3 - StructuredGrid.plot_curvature() works fine

Important to say that `StructuredGrid.plot_curvature()` performs well in all cases:
- Jupyter Notebook
- VS Code
- pyvista-examples and my own data (`my_pv_grid.vtk`) 
![image](https://user-images.githubusercontent.com/9113903/87443097-6dd61100-c5cb-11ea-815a-e3b985f0f299.png)

## Some specs
Windows 10
pyvista version 0.25.3

## What do do?
Has anyone encountered this problem already? How to solve?

Thanks! 😄

**Comment by banesullivan** [comment]
Okay, so most of these are all pretty simple issues. Let me know if anything is unclear

## 1

> However, I get different results for the same parts of the code. VS Code makes it white, while Jupyter Notebook makes it colored. The piece of code that corresponds to the images is:

The different coloring scheme has to do with the active plotting theme.

This has to do with the fact that we add a line to the top of the notebooks when you download the notebook from our website. You'll notice that the notebook has the following code at the top which isn't in the downloaded `.py` file. Simply copy this code snippet to the `.py` to get the same result:

```py
from pyvista import set_plot_theme
set_plot_theme('document')
```

> It seems there something about StructuredGrid.plot() in VS Code. Whether I pass or don't cmap as an argument, it does not make any difference. I tried two different environments:

`cmap` will not matter here because there are no scalar data arrays to map colors to. YOu can only plot that mesh as a solid color. Use the `color` argument to control that. When you add data to the mesh later in the example, you can plot those data with a colormap

## 2

Again with the plotting theme and lack of scalar data. The mesh you shared has no scalar array to color map. Try using a filter in PyVista like `.elevation` to add a scalar array:

```py
import pyvista as pv

# read StructuredGrid data
pvgrid = pv.read('my_pv_grid.vtk')

with_data = pvgrid.elevation()

with_data .plot()
```

## 3

Expected because `plot_curvature` adds a `curvature` scalar array to map colors to.

------

So in summary: set the plot theme with the code snippet above and realize that you cannot map colors to the mesh without scalar data to map to.

**Comment by leomiquelutti** [comment]
Hi @banesullivan thanks for the amazingly fast response! 😄 

So, AFAIU, although I created a `StructuredGrid` from three 2D numpy-arrays as below, I didn't assign "measured" values to them. As you can tell, I created a `StructuredGrid` of magnetic data.

```
# grid.TMI is an xarray obtained using Verde from Fatiando
x = np.tile(grid.TMI.easting.values, (len(grid.TMI.northing.values),1))
y = np.tile(grid.TMI.northing.values, (len(grid.TMI.easting.values),1))
pvgrid = pv.StructuredGrid(x, np.transpose(y), grid.TMI.data)
```

To add scalar values to it, I only call `pvgrid = pvgrid.elevation()` and the assignment is complete, right?

Other questions arise:

1. Is there another way to add these scalar values?  Or `elevation()` would be a _default_ way?
2. Am I using the correct class for my purpose, i.e. `StructuredGrid`? Initially, my purpose was simply to create a map of the TMI values.
3. Is it possible to create a flat image, as a regular map?
4. Is there another straightforward way to save it as an image than plot(screenshot='foo.png')?
5. Should I ask these new questions here or in a new topic? Or in Slack? If you prefer for the sake of organization I can create a new topic.

**Comment by banesullivan** [comment]
> So, AFAIU, although I created a StructuredGrid from three 2D numpy-arrays as below, I didn't assign "measured" values to them. As you can tell, I created a StructuredGrid of magnetic data.

I'd recommend looking over the introduction to PyVista's data structures where we explain the role of meshes and how you can associate scalar/vector data to the mesh. See https://docs.pyvista.org/getting-started/what-is-a-mesh.html

Otherwise, yep. You created a bare mesh and did not add any point/cell arrays to the mesh so there wasn't any data to map.

> To add scalar values to it, I only call pvgrid = pvgrid.elevation() and the assignment is complete, right?

This is one filter (there are [many filters](https://docs.pyvista.org/core/filters.html)) that adds scalar data to the mesh and happened to be what you were hoping for. This filter ad a scalar array called `'Elevation'` to the mesh. You can access that array by:

```py
>>> pvgrid['Elevation']
```

> Is there another way to add these scalar values? Or elevation() would be a default way?

You can add any point/cell arrays to the mesh via bracket indexing with the array name. For example, add som point or cell ids to the mesh:

```py
pvgrid['pids'] = np.arange(pvgrid.n_points)

pvgrid['p=cids'] = np.arange(pvgrid.n_cells)
```

or you could add your data to the mesh via:

```py
pvgrid['data'] = grid.TMI.data.ravel() # you may need to ravel with `order=F`
```

> Am I using the correct class for my purpose, i.e. StructuredGrid? Initially, my purpose was simply to create a map of the TMI values.

Yep! `StructuredGrid` is a good class for your particular data.

> Is it possible to create a flat image, as a regular map?

You can make the mesh flat by passing in zeros (or anther constant) in the z-component when initializing the `StrucutredGrid`. E.g.

```py
pvgrid = pv.StructuredGrid(x, np.transpose(y), np.zeros_like(grid.TMI.data))
```

If you're looking for 2D plotting, Matplotlib is going to be better suited for you.

> Is there another straightforward way to save it as an image than plot(screenshot='foo.png')?

What do you mean? A screenshot of the scene? if so. that's the way to do it.

> Should I ask these new questions here or in a new topic? Or in Slack? If you prefer for the sake of organization I can create a new topic.

Usage questions like this need to be on GitHub, not slack!! For the moment, these questions are all fine to bundle here because they are all general use/beginner questions. As you have more questions that get more specific, please open a new issue/thread.

Otherwise, I'd highly recommend skimming through the [What is a Mesh?](https://docs.pyvista.org/getting-started/what-is-a-mesh.html) and [Basic API Usage](https://docs.pyvista.org/getting-started/simple.html) sections of the documentation, then of course continuing with more demos from [the gallery](https://docs.pyvista.org/examples/index.html)

**Comment by leomiquelutti** [comment]
Thanks again @banesullivan!

I've been through the docs for hours already; and that didn't prevent me from misunderstanding some basics 😂. I try to be a good boy before asking for your (or any developer's) time 😉

I believe that one thing that bottlenecked me was my misconception about the usage, because in the end whether yours is a 2D grid or a 3D mesh, as you say in the docs, `pyvista` always consider them 3D. And that was fortunate because it gave me the chance of look at a geophysical map (tmi) with a new perspective! 

About the _screenshot_, I meant something that I could export to georeference in a Globe (as Google Earth, or Cesium, that requires the image and its edges coordinates). That means a _jpg_ with no borders, colorbar, and similars. `plot(screenshot='foo.png')` gets the whole scene. But I don't know if what I need makes sense when we talk about 3D objetcs...

**Comment by banesullivan** [comment]
> I've been through the docs for hours already; and that didn't prevent me from misunderstanding some basics 😂. I try to be a good boy before asking for your (or any developer's) time 😉

If anything is unclear in the documentation, please open new issues to help us refine the documentation so that it is more understandable/approachable to new users. We (the developers of PyVista) write those docs and sometimes we lose perspective and assume the reader knows more than we should expect them to.

> I believe that one thing that bottlenecked me was my misconception about the usage, because in the end whether yours is a 2D grid or a 3D mesh, as you say in the docs, pyvista always consider them 3D. And that was fortunate because it gave me the chance of look at a geophysical map (tmi) with a new perspective!

Ah cool! Yep, all meshes in PyVista have a complete 3D spatial reference. After all, any geospatial/2D data are 3D as well, they just have a constant elevation or wrap to topography.

> About the screenshot, I meant something that I could export to georeference in a Globe (as Google Earth, or Cesium, that requires the image and its edges coordinates). That means a jpg with no borders, colorbar, and similars. plot(screenshot='foo.png') gets the whole scene. But I don't know if what I need makes sense when we talk about 3D objetcs...

This isn't really the best way to go about that... You'll want to export the mesh to a format that Cesium accepts (please open a new issue for that and I can share some code on how to do that there) then load that mesh into cesium. While I can help you export a mesh from PyVista to a format like glTF/glb, I cannot help you do anything with that in Cesium.

**Comment by leomiquelutti** [comment]
> If anything is unclear in the documentation, please open new issues to help us refine the documentation so that it is more understandable/approachable to new users. We (the developers of PyVista) write those docs and sometimes we lose perspective and assume the reader knows more than we should expect them to.

I believe that the nature of pyvista per se poses some challenges for users, especially beginners. I am no expert at all in writing code documentation, but I believe that docs affect users differently, according to the user's way of learning things. For example, to me, the fact that examples are always loaded (from pyvista import examples) and we don't see the creation of the objects is a disadvantage because I usually run and debug examples after some doc skimming, but usually these two (reading the docs and running examples) are concomitant. Since pyvista has so many classes and is deeply linked to vtk classes made it a little bit worse. 

So far I will not open an issue about this because I don't have anything specific to point out about the docs but my personal experience with the learning-pyvista process. If one day any of you (pyvista developers) feel like you want to talk about this, feel free to contact me and I will be more than happy to provide you any feedback you may want.

> This isn't really the best way to go about that... You'll want to export the mesh to a format that Cesium accepts (please open a new issue for that and I can share some code on how to do that there) then load that mesh into cesium. While I can help you export a mesh from PyVista to a format like glTF/glb, I cannot help you do anything with that in Cesium.

I'll (re)think of our approach with images/models and Cesium and probably soon will open a new issue about it. Thanks a lot @banesullivan! 🙌
